chain/gap/find.go

```diff
@@ -0,0 +1,264 @@
+		Where("height >= ?", min).
+		Where("height <= ?", max).
+		Select(); err != nil {
+		return nil, xerrors.Errorf("query processing report skips: %w", err)
+	}
+	log.Debugw("executed find skipped epoch query", "count", len(skippedReports))
+
+	gapReport := make([]*visor.GapReport, len(skippedReports))
+	for idx, r := range skippedReports {
+		gapReport[idx] = &visor.GapReport{
+			Height:     r.Height,
+			Task:       r.Task,
+			Status:     "GAP",
+			Reporter:   g.name,
+			ReportedAt: reportTime,
+		}
+	}
+	return gapReport, nil
+}
+
+func (g *GapIndexer) findEpochGapsAndNullRounds(ctx context.Context, node GapIndexerLens, max, min uint64) (visor.GapReportList, []abi.ChainEpoch, error) {
+	log.Debug("finding epoch gaps and null rounds")
+	reportTime := time.Now()
+
+	var nullRounds []abi.ChainEpoch
+	var missingHeights []uint64
+	res, err := g.DB.AsORM().QueryContext(
+		ctx,
+		&missingHeights,
+		`
+		SELECT s.i AS missing_epoch
+		FROM generate_series(?, ?) s(i)
+		WHERE NOT EXISTS (SELECT 1 FROM visor_processing_reports WHERE height = s.i);
+		`,
+		min, max)
+	if err != nil {
+		return nil, nil, err
+	}
+	log.Debugw("executed find epoch gap query", "count", res.RowsReturned())
+
+	gapReport := make([]*visor.GapReport, 0, len(missingHeights))
+	// walk the possible gaps and query lotus to determine if gap was a null round or missed epoch.
+	for _, gap := range missingHeights {
+		gh := abi.ChainEpoch(gap)
+		tsgap, err := node.ChainGetTipSetByHeight(ctx, gh, types.EmptyTSK)
+		if err != nil {
+			return nil, nil, xerrors.Errorf("getting tipset by height %d: %w", gh, err)
+		}
+		if tsgap.Height() == gh {
+			log.Debugw("found gap", "height", gh)
+			for _, task := range chain.AllTasks {
+				gapReport = append(gapReport, &visor.GapReport{
+					Height:     int64(tsgap.Height()),
+					Task:       task,
+					Status:     "GAP",
+					Reporter:   g.name,
+					ReportedAt: reportTime,
+				})
+			}
+		} else {
+			log.Debugw("found null round", "height", gh)
+			nullRounds = append(nullRounds, gh)
+		}
+	}
+	return gapReport, nullRounds, nil
+}
+
+type TaskHeight struct {
+	Task   string
+	Height uint64
+	Status string
+}
+
+// TODO rather than use the length of `tasks` to determine where gaps are, use the contents to look for
+// gaps in specific task. Forrest' SQL-Foo isn't good enough for this yet.
+func (g *GapIndexer) findTaskEpochGaps(ctx context.Context, max, min uint64, tasks ...string) (visor.GapReportList, error) {
+	log.Debug("finding task epoch gaps")
+	start := time.Now()
+	var result []TaskHeight
+	var out visor.GapReportList
+	// returns a list of tasks and heights for all incomplete heights
+	// and incomplete height is a height with less than len(tasks) entries, the tasks returned
+	// are the completed tasks for the height, we can diff them against all known tasks to find the
+	// missing ones.
+	res, err := g.DB.AsORM().QueryContext(
+		ctx,
+		&result,
+		`
+select height, task
```

Comment:
I ran this against mainnet on range 900_000 to 1_000_000 and it took about 3 seconds fwiw.